pydantic/env_settings.py

```diff
@@ -44,19 +45,58 @@ def _build_values(
         _env_file_encoding: Optional[str] = None,
         _secrets_dir: Union[Path, str, None] = None,
     ) -> Dict[str, Any]:
+        extra_settings = [source(self) for source in reversed(self.__config__.extra_settings_sources)]
         return deep_update(
-            self._build_secrets_files(_secrets_dir), self._build_environ(_env_file, _env_file_encoding), init_kwargs
+            *extra_settings,
+            self._build_secrets_files(_secrets_dir),
+            self._build_environ(_env_file, _env_file_encoding),
+            init_kwargs,
         )
 
+    def filter_relevant_env_vars(self, env_vars: Mapping[str, Optional[str]]) -> Dict[str, Optional[str]]:
+        """
+        Helper function to keep only relevant settings vars from an external source.
+        """
+        d: Dict[str, Optional[str]] = {}
+
+        for field in self.__fields__.values():
+            env_val: Optional[str] = None
+            for env_name in field.field_info.extra['env_names']:
+                env_val = env_vars.get(env_name)
+                if env_val is not None:
+                    break
+
+            if env_val is None:
+                continue
+
+            if field.is_complex():
+                try:
+                    env_val = self.__config__.json_loads(env_val)  # type: ignore
+                except ValueError as e:
+                    raise SettingsError(f'error parsing JSON for "{env_name}"') from e
+            d[field.alias] = env_val
+
+        return d
+
+    def load_env_vars_from_source(self, loader: Callable[[str], Any]) -> Dict[str, Optional[str]]:
+        """
+        Helper function to load all relevant fields from an external source using a provided loader.
+        """
+        d: Dict[str, Optional[str]] = {}
+        for field in self.__fields__.values():
+            for env_name in field.field_info.extra['env_names']:
+                value = loader(env_name)
+                if value != undefined:
```

Comment:
surely we should use `None` or `KeyError` here?

Reply:
I would prefer to keep the `undefined` as it allows the source to support any type including `None`. 
About the `KeyError` check, I think it's too specific to dict based sources: in the case of a REST API based source, a call to `loader` can trigger an API call that simply returns a flat value.
What do you think ?